acceptance/acceptance_test.go

```diff
@@ -229,1905 +99,1417 @@ func testWithoutSpecificBuilderRequirement(
+					"-b", builderConfigPath,
+				)
+
+				assert.NotNil(err)
+				assert.NewOutputAssertionManager(output).ReportsWindowsContainersExperimental()
+			})
+		})
+
+		when("experimental is enabled", func() {
+			it("succeeds", func() {
+				stackManager.EnsureDefaultStackCreated()
+
+				createBuilderPack.EnableExperimental()
+
+				testBuilder := builderManager.CreateWindowsBuilder()
+				defer testBuilder.Cleanup()
+
+				assert.ImageIsWindows(testBuilder.Name())
+			})
+		})
 	})
 
-	when("stack is created", func() {
+	when("build", func() {
 		var (
-			runImageMirror string
+			assertAppImage          assertions.ImageAssertionManager
+			appName, appImageName   string
+			mockAppPath             = filepath.Join("testdata", "mock_app")
+			mixedComponentMessenger components.MixedComponents
+			cacheCleanupRequired    bool
 		)
 
 		it.Before(func() {
-			value, err := suiteManager.RunTaskOnceString("create-stack",
-				func() (string, error) {
-					runImageMirror := registryConfig.RepoName(runImage)
-					err := createStack(t, dockerCli, runImageMirror)
-					if err != nil {
-						return "", err
-					}
-
-					return runImageMirror, nil
-				})
-			h.AssertNil(t, err)
+			cacheCleanupRequired = false
+			h.SkipIf(t, containerManager.HostOS() == "windows", "These tests are not yet compatible with Windows-based containers")
 
-			suiteManager.RegisterCleanUp("remove-stack-images", func() error {
-				return h.DockerRmi(dockerCli, runImage, buildImage, value)
-			})
+			appName = fmt.Sprintf("some-org/%s", h.RandString(10))
+			appImageName = registry.RepoName(appName)
+
+			assertAppImage = assert.NewImageAssertionManager(appImageName, containerManager)
+
+			mixedComponentMessenger = components.NewMixedComponents(t, testLifecycle, pack)
+
+			cacheCleanupRequired = true
+		})
+
+		it.After(func() {
+			if cacheCleanupRequired {
+				h.DockerRmi(dockerCli, appImageName)
+				ref, err := name.ParseReference(appImageName, name.WeakValidation)
+				assert.Nil(err)
+				cacheImage := cache.NewImageCache(ref, dockerCli)
+				buildCacheVolume := cache.NewVolumeCache(ref, "build", dockerCli)
+				launchCacheVolume := cache.NewVolumeCache(ref, "launch", dockerCli)
+				cacheImage.Clear(context.TODO())
+				buildCacheVolume.Clear(context.TODO())
+				launchCacheVolume.Clear(context.TODO())
+			}
+		})
+
+		it("creates a runnable, rebuildable image on daemon from app dir", func() {
+			stackManager.EnsureDefaultStackCreated()
+			testBuilder := builderManager.EnsureComboBuilderExists()
+			pack.SuccessfulRun("set-default-builder", testBuilder.Name())
+			if pack.Supports("trust-builder") {
+				pack.SuccessfulRun("trust-builder", testBuilder.Name())
+			}
+
+			appPath := filepath.Join("testdata", "mock_app")
+			output := pack.SuccessfulRunWithOutput("build", appImageName, "-p", appPath)
+			cleanupAppImageTask := containerManager.CleanupTaskForImageByName(appImageName)
+			defer cleanupAppImageTask()
```

Comment:
Is there a difference between running the cleanup as part of `it.After...` and setting it under `defer`?
Sometimes you choose the first option and sometimes the second one.
I'm not sure what is better.
You'll probably less likely forget to add the `defer` but I think the code is clearer when all of the "after commands" appear together.

Reply:
On second thought, it's clear, in this case, why you have the `defer` and not the `it.After` but my question is not specifically to this case.

Reply:
@dfreilich had concerns around a similar area, I [commented](https://github.com/buildpacks/pack/pull/673#discussion_r442912727) on his question.

That doesn't cover the `After()` vs `defer` discussion though. If we perform an action in an `After` it will always run, regardless of how much of the test was executed, which could cause further failures and pollute our test output. With a `defer` we can tightly couple to the stage of the test that instantiates an object. We also potentially don't have to share a context wide state - I'd rather avoid an additional nested context in which we'd define some variables and an `After` if I can.

That said, `defer` does come with downsides, failures that occurs in a `defer` are unreliably reported to the tester.

I think the differences throughout the code are likely to be a combination of different circumstances feeling more or less appropriate, influence from the existing test suite and how I felt as they were written. I don't know if any of them a right or wrong.

Reply:
Thank you for your detailed answer. It totally makes sense!